include/picongpu/plugins/randomizedParticleMerger/VoronoiCell.hpp

```diff
@@ -0,0 +1,367 @@
+            }
+
+            const float3_X momentum2 = momentum * momentum;
+
+            for( int i = 0; i < DIM3; i++ )
+            {
+                cupla::atomicAdd( acc, &this->meanMomentumValue[i], weighting * momentum[i], ::alpaka::hierarchy::Threads{} );
+                cupla::atomicAdd( acc, &this->meanMomentumSquaredValue[i], weighting * momentum2[i], ::alpaka::hierarchy::Threads{} );
+            }
+        }
+
+        /** Counting parameters that are necessary before processing vornoi cell:
+         *  mean values and expected number of macro particles
+         *
+         * @param minMacroParticlesToDivide min number of macroparticles in a cell
+         *                                  such that the cell is always subdivided
+         * @param ratioKeptParticles ratio of particles that are kept on average
+         */
+        HDINLINE
+        void finalizePrecalculationValues(
+            const uint32_t minMacroParticlesToDivide,
+            const float_X ratioKeptParticles
+        )
+        {
+            finalizeMeanValues();
+            finalizeExpectedNumberParticles( minMacroParticlesToDivide, ratioKeptParticles );
+
+        }
+
+        //! Finalize calculation of mean values
+        HDINLINE
+        void finalizeMeanValues()
+        {
+            meanMomentumValue /= numRealParticles;
+            meanPositionValue /= numRealParticles;
+            meanMomentumSquaredValue /= numRealParticles;
+            meanPositionSquaredValue /= numRealParticles;
+        }
+
+        /** Count expected number of particles in the cell
+         *
+         * @param minMacroParticlesToDivide min number of macroparticles in a cell
+         *                                  such that the cell is always subdivided
+         * @param ratioKeptParticles ratio of particles that are kept on average
+         */
+        HDINLINE
+        void finalizeExpectedNumberParticles(
+            uint32_t const minMacroParticlesToDivide,
+            float_X const ratioKeptParticles
+        )
+        {
+            // Special case for the original voronoi cells
+            if ( parentExpectedNumMacroParticles < 0 )
+            {
+                expectedNumMacroParticles = numMacroParticles * ratioKeptParticles;
+                return;
+            }
+
+            // Algorithm stop conditions for 1 and 2 macroparticles
+            if ( numMacroParticles == 1u )
+                expectedNumMacroParticles = 1.0_X;
+            if ( numMacroParticles == 2u && parentNumMacroParticles == 3u )
+                expectedNumMacroParticles = 2.0_X;
+
+            // Normal subdivision step
+            if ( parentNumMacroParticles > minMacroParticlesToDivide )
+            {
+                expectedNumMacroParticles = numMacroParticles * ratioKeptParticles;
+            }
+            else
+            {
+                // TODO: this equation works, but there may be a better one (see the notebook)
```

Comment:
should be `@todo` 

I am not sure if someone expect you can follow the comment because we do not now the notebook.
suggestion: move the equation to gist and link it here, or copy the equation to the todo

Reply:
It's not just one equation to be changed. I think we could still remove this comment.